add verification key endpoint

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
https://jira.iohk.io/browse/ADP-474

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have added endpoint plus basic types and instances
- [x] I have extended swagger
- [x] I have updated core unit tests
- [x] I have added impl and switched on impl in Shelley
- [x] I have extended Api.Link and core-integration's DSL
- [x] I have added illustrative integration test
- [x] I have added golden tests for explicit mnemonic and picked indices based on cardano-addresses

# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

rvl commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -804,6 +810,19 @@ newtype MinWithdrawal = MinWithdrawal
+instance FromHttpApiData ApiRelativeAddressIndex where
+    parseUrlPiece = bimap (T.pack . getTextDecodingError) ApiRelativeAddressIndex . fromText
+
+instance FromText Word31 where
```
This could go into the text-class package I think, and perhaps it could be defined in terms of another typeclass like Integral.

paweljakubas replied:
done, although now it is Word32. After the rebase many things need to change

rvl reviewed (commented):
Looking good

paweljakubas committed:
add endpoint scaffolding plus needed types

paweljakubas committed:
update swagger

paweljakubas committed:
add core unit testing

paweljakubas committed:
add impl of getting key hash using cardano-addresses

Commit:
Regenerate nix

paweljakubas committed:
add getWalletKey to api link

paweljakubas committed:
extend DSL

paweljakubas committed:
add integration test

paweljakubas committed:
add golden test based on cardano-addresses

KtorZ commented on lib/core/src/Cardano/Wallet/Api.hs:
```diff
@@ -273,6 +279,22 @@ type GetUTxOsStatistics = "wallets"
+type GetWalletKey = "wallets"
+    :> Capture "walletId" (ApiT WalletId)
+    :> "keys"
+    :> "script"
```
What about capturing the full role here, so that we get better re-usability for this endpoint.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1101,6 +1108,26 @@ getUTxOsStatistics ctx (ApiT wid) = do
+        let (KeyHash bytes) =
+                hashKey $ deriveMultisigPublicKey
+                (liftXPub $ Shelley.getKey accPub)
+                (toEnum $ fromInteger $ toInteger index)
```
That logic would better be located in `Cardano.Wallet.hs`. What about defining a function along those lines:

```hs
getWalletKeyHash 
    ::  ctx
    -> WalletId
    -> AccountingStyle
    -> Index 'Soft AddressK
    -> KeyHash
```

and re-use it in this layer?

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -711,6 +712,10 @@ data ApiWalletMigrationInfo = ApiWalletMigrationInfo
 
+newtype ApiVerificationKeyHash = ApiVerificationKeyHash
+    { verificationKeyHash :: ApiT (Hash "ScriptKey")
+    } deriving (Eq, Generic, Show)
```
Alternatively, we could have the API return a plain bech32-encoded string on success?

KtorZ commented on lib/text-class/src/Data/Text/Class.hs:
```diff
@@ -116,6 +118,19 @@ instance FromText Natural where
 
+instance FromText Word32 where
+    fromText =
+        validate <=< (fmap fromIntegral . fromText @Natural)
```
This will not work :scream:! The `fromIntegral` that transform from `Natural` to `Word32` needs to happen _after_ the validation to be safe. Otherwise, this may just throw if the value can't be converted back. 

Note that @hasufell is doing something similar but re-using constructor from the `text` package which makes things a bit nicer: https://github.com/input-output-hk/cardano-wallet/pull/2213/files#diff-b892068f732dfe3e67cd6f7a464975ba21a3588de3e5ea74022be5dfaa18e476

I'd suggest cherry-picking this change :+1:

KtorZ commented on stack.yaml:
```diff
@@ -44,9 +44,9 @@ extra-deps:
+# cardano-addresses
 - git: https://github.com/input-output-hk/cardano-addresses
-  commit: e0ab4587266430a08734e1aec1c29d261a9a3b70
+  commit: 02fb4ec7db7c55334ab2216c94cf70b209a33885
```
We need to use released version of cardano-addresses. There are some cleanup to be done in cardano-addresses before we can release a 3.0.0 version. Let's discuss that on Slack.

KtorZ committed:
wire cables for arbitrary message signing in the API.

KtorZ committed:
revert cardano-addresses bump. Sticking to last release for now.

KtorZ committed:
generalize public key derivation in the API for all roles.

  And also, to keys and not key hashes. The goal of this endpoint is to
  expose public keys so clients can perform signature verification all
  by themselves.

KtorZ committed:
make sign metadata endpoint consistent with others (move 'signatures' after wallet id)

KtorZ committed:
add integration tests for signatures produced by the API

  In particular, we use in conjunction with the other endpoint to get a public key, and show that the public key can verify the signature.

Commit:
fixup! Regenerate nix
